revokefs/writer.c

```diff
@@ -0,0 +1,883 @@
+{
+  return request_path_i64_i64 (writer_socket, REVOKE_FS_TRUNCATE, path, size, 0);
+}
+
+static ssize_t
+handle_utimens (RevokefsRequest *request,
+                gsize data_size,
+                RevokefsResponse *response)
+{
+  g_autofree char *path = NULL;
+  struct timespec *tv;
+
+  if (request->arg1 + sizeof (struct timespec) * 2 != data_size)
+    {
+      g_printerr ("Invalid data size\n");
+      exit (1);
+    }
+
+  path = get_valid_path (request->data, request->arg1);
+  tv = (struct timespec *)(request->data + request->arg1);
+
+  if (utimensat (basefd, path, tv, AT_SYMLINK_NOFOLLOW) == -1)
+    response->result = -errno;
+  else
+    response->result = 0;
+
+  return 0;
+}
+
+int
+request_utimens (int writer_socket, const char *path, const struct timespec tv[2])
+{
+  return request_path_data (writer_socket, REVOKE_FS_UTIMENS, path,
+                            (guchar *)tv, sizeof (struct timespec) * 2);
+}
+
+static ssize_t
+handle_open (RevokefsRequest *request,
+             gsize data_size,
+             RevokefsResponse *response)
+{
+  g_autofree char *path = get_valid_path (request->data, data_size);
+  int mode = request->arg1;
+  int flags = request->arg2;
+  int fd;
+
+
+  /* We need to specially handle O_TRUNC */
+  fd = openat (basefd, path, flags & ~O_TRUNC, mask_mode (mode));
+  if (fd == -1)
+    response->result = -errno;
+  else
+    {
+      response->result = 0;
+      if (flags & O_TRUNC)
+        {
+          if (ftruncate (fd, 0) == -1)
+            response->result = -errno;
+        }
+
+      if (response->result == 0)
+        {
+          g_hash_table_insert (outstanding_fds, GUINT_TO_POINTER(fd), GUINT_TO_POINTER(1));
+          response->result = fd;
+        }
+      else
+        (void) close (fd);
+    }
+
+  return 0;
+}
+
+int
+request_open (int writer_socket, const char *path, mode_t mode, int flags)
+{
+  return request_path_int_int (writer_socket, REVOKE_FS_OPEN, path, mode, flags);
+}
+
+static ssize_t
+handle_read (RevokefsRequest *request,
+             gsize data_size,
+             RevokefsResponse *response)
+{
+  int r;
+  int fd = request->arg1;
+  size_t size = request->arg2;
+  off_t offset = request->arg3;
+
+  if (size > MAX_DATA_SIZE)
+    size = MAX_DATA_SIZE;
+
+  if (g_hash_table_lookup (outstanding_fds, GUINT_TO_POINTER(fd)) == NULL)
+    {
+      response->result = -EBADFD;
+      return 0;
+    }
+
+  r = pread (fd, response->data, size, offset);
+  if (r == -1)
+    {
+      response->result = -errno;
+      return 0;
+    }
+  else
+    {
+      response->result = r;
+      return r;
+    }
+}
+
+int
+request_read (int writer_socket, int fd, char *buf, size_t size, off_t offset)
+{
+  RevokefsRequest request = { REVOKE_FS_READ };
+  RevokefsResponse response;
+  ssize_t response_data_len;
+
+  request.arg1 = fd;
+  request.arg2 = size;
+  request.arg3 = offset;
+
+  response_data_len = do_request (writer_socket, &request, NULL, 0, NULL, 0,
```

Comment:
In support of the idea that `do_request` should do the locking itself, the locking is missing here.

Reply:
I thought we want to keep read requests multi-threaded. That's why I didn't lock `do_request` entirely instead lock all write-related ops. Sometime ago from IRC:

```
<wjt> uajain alexlarsson: I've found a bit of a problem with this revokefs prototype: by default, fuse uses a multi-threaded mainloop, and there's no serialization of requests/responses exchanged with the writer process
<alexlarsson> wjt: ah, well, just disable the fuse multitheading?
<wjt> uajain: try passing `"-s"` to the fuse process – I think it'll work much better :)
<alexlarsson> or serialize it
<wjt> right. I guess only access to the socket needs to be serialized; all the read-only stuff can happily happen in 8 threads
<alexlarsson> It might be useful to be multithreaded on the reader side
<wjt> maybe holding a lock for the duration of do_request() would be enough
```

Reply:
This means filesystem reads, which are handled entirely in the fuse process. Reads on the socket must be serialised because otherwise you could have this case:

* FUSE thread A writes a request P to the socket
* FUSE thread B writes a request Q to the socket
* Writer process responds to P
* Writer process responds to Q
* FUSE thread B incorrectly reads the response to P